hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -842,16 +845,55 @@ public void startCommitWithTime(String commitTime) {
         new HoodieInstant(true, commitActionType, commitTime));
   }
 
+  /**
+   * Provides a new commit time for a compaction (commit) operation
+   */
+  public String startCompaction() {
+    String commitTime = HoodieActiveTimeline.createNewCommitTime();
+    logger.info("Generate a new commit time " + commitTime);
+    startCompactionWithTime(commitTime);
+    return commitTime;
+  }
+
+  /** Since MOR tableType default to {@link HoodieTimeline#DELTA_COMMIT_ACTION},
+   * we need to explicitly set to {@link HoodieTimeline#COMMIT_ACTION} for compaction
+   */
+  public void startCompactionWithTime(String commitTime) {
+    HoodieTable<T> table = HoodieTable.getHoodieTable(
+        new HoodieTableMetaClient(jsc.hadoopConfiguration(), config.getBasePath(), true),
+        config);
+    HoodieActiveTimeline activeTimeline = table.getActiveTimeline();
+    String commitActionType =  HoodieTimeline.COMMIT_ACTION;
+    activeTimeline.createInflight(
+        new HoodieInstant(true, commitActionType, commitTime));
+  }
+
+  /**
+   * Performs a compaction operation on a dataset. WARNING: Compaction operation cannot be executed
+   * asynchronously. Please always use this serially before or after an insert/upsert action.
+   */
+  public JavaRDD<WriteStatus> compact(String commitTime) throws IOException {
+    // Create a Hoodie table which encapsulated the commits and files visible
+    HoodieTable<T> table = HoodieTable.getHoodieTable(
+        new HoodieTableMetaClient(jsc.hadoopConfiguration(), config.getBasePath(), true),
+        config);
+    List<CompactionOperation> operations = ((HoodieMergeOnReadTable) table).getCompactionWorkload(jsc, config, commitTime);
```

Comment:
can we avoid the type casting and make it a no-op for COW table? i.e operations can be empty..

Reply:
This means will have to add getCompactionWorkload() needs to be added to HoodieTable interface...

Reply:
if we need a little more time with API design, we can discuss this. But I don't think we should cast table types in HoodieWriteClient level..